frame/contracts/common/src/lib.rs

```diff
@@ -37,18 +37,18 @@ use serde::{Serialize, Deserialize};
 pub struct ContractResult<T> {
 	/// How much gas was consumed during execution.
 	pub gas_consumed: u64,
-	/// An optional debug message. This message is only non-empty when explicitly requested
-	/// by the code that calls into the contract.
+	/// An optional debug message. This message is only filled when explicitly requested
+	/// by the code that calls into the contract. Otherwise it is empty.
 	///
 	/// The contained bytes are valid UTF-8. This is not declared as `String` because
-	/// this type is not allowed within the runtime. A client should decode them in order
-	/// to present the message to its users.
+	/// this type is not allowed within the runtime.
 	///
 	/// # Note
 	///
 	/// The debug message is never generated during on-chain execution. It is reserved for
 	/// RPC calls.
-	pub debug_message: Bytes,
+	#[cfg_attr(feature = "std", serde(with = "as_string"))]
+	pub debug_message: Vec<u8>,
```

Comment:
Just had another thought - what about making this a `debug_messages: Vec<Vec<u8>>`? Otherwise a UI client would not know how to display messages discretely without resorting to a convention relying on a custom separator e.g. `\n`.

Reply:
That is a good suggestion. Current the API decides this, but especially if we get traces, it will be as messy as the RPC error results.

Reply:
Clients should not make any assumptions about the format of the buffer. They should just display it as-is. The reason is that it will not only contain log messages but also information added by the pallet. Like detailed information about errors and a call stack trace.

This is also why I renamed from `seal_println` to `seal_debug_message`. It is not necessarily a line. It just appends to the buffer and allows contracts to create a line using multiple calls.